GenericList/string_list.h

```diff
@@ -1,31 +1,78 @@
 #pragma once
 
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <memory>
+#include <cassert>
+
 class List_String {
+    //note: elements are value copies of original objects (copy-by-value, not copy-by-reference)
 public:
     List_String();
-    ~List_String();
 
-    int Capacity();
-    bool Capacity(int new_capacity);
-    int Count();
+    friend void swap(List_String& first, List_String& second);
+
+    //Rule of 3
+    ~List_String(); //Destructor
+    List_String(const List_String& other); //Copy Constructor
+    List_String& operator=(const List_String& other); //Copy assignement
+
+    //Rule of 5
+    //R-value references (&&) explained: http://thbecker.net/articles/rvalue_references/section_01.html 
+    //basic explanation: if a function argument is &&, whatever it references will stop existing at the end of the function
+    List_String(List_String&& other);
+    List_String& operator=(List_String&& other);
+
+
+
+
+    int Capacity() const;
+    int Count() const;
+
+    //Sets the capacity of the internal array to new_capacity. If new_capacity is smaller than Count, do nothing.
+    void Capacity(int new_capacity);
+
+    std::string ToString() const;
 
-    std::string ToString();
+    void Add(const std::string& new_val);
+    void Add(std::string&& new_val);
 
-    bool Add(std::string new_val);
-    bool Contains(std::string val);
+    bool Contains(const std::string& val) const;
     bool RemoveAt(int index);
+
+    int IndexOf(const std::string& val) const;
+
+    //Returns true if a relevant element exists, and removes it.
+    bool Remove(const std::string& val);
```

Comment:
You can subsequently add `Find` and `Remove` overloads that takes a projector, e.g.:
```
template<typename Arg, typename Project>
const std::string* Find(const Arg& arg, Project project) const {
    for (const auto& val : *this) {
        if (val == project(arg)) return &val;
    }
    
    return nullptr;
}
```
Usage would be something like this:
```
struct AbstractClass {
    ~AbstractClass() = default:
    virtual const char* GetName() const = 0;
}

AbstactClass* ptr;

// find a string that matches the name of an abstract class.
list.Find(ptr, [](AbstractClass* p) { return p->GetName(); });
```
Look up lambda expressions in C++

Reply:
PS The standard library does not implement these in `std::vector` because the generic functions in `<algorithm>` works for most if not all containers defined by stl.

Reply:
You can subsequently add `Find` and `Remove` overloads that take a predicate, e.g.:
```
template<typename Predicate>
const std::string* Find(Predicate pred) const {
    for (const auto& val : *this) {
        if (pred(val)) return &val;
    }
    
    return nullptr;
}
```
Usage would be something like this:
```
struct AbstractClass {
    ~AbstractClass() = default:
    virtual const char* GetName() const = 0;
}

AbstactClass* ptr;

//Find the first string where the predicate evaluates to true
list.Find([ptr](const std::string& val) { return val == ptr->GetName(); });
```
Look up lambda expressions in C++

Reply:
One issue with using Predicates using templates is it causes compiler errors when calling Find with arguments that can be cast to std::string (but that aren't std::string). An example is `Find("text")`, which will attempt to call the predicate template version of Find instead of `Find(std::string)`.
Is the best way to work around this using a clear indicate that we want unary functions?
ie: `std::string* Find(std::function<bool<std::string>> f);`

Reply:
You could've just renamed the function with the predicate to `FindIf`. One issue that stands out currently is that it should be `std::function<bool(const std::string&)>`, you are copying each element otherwise. 

Beware of the hidden costs in `std::function` though, if the object used to construct the `std::function` instance is larger than a certain size (typically 32 bytes) it's going to do a heap allocation. 

If it's a function I'm exporting, i.e. callable across dynamic libraries, I would _personally_ prefer the C-style function pointer with `void*` context data, but would use `std::function` if the team I'm working with prefers it, which is typically the case. If it's an internal function or a function I can define completely in the header I would usually write the template version I had before. There is a way around the compile error you mentioned, it's fairly advanced C++:

```
#include <type_traits>

template<typename Predicate>
std::enable_if_t<
    std::is_invocable_v<Predicate, std::string>,
const std::string*> Find(Predicate&& pred) const { for (const auto& val : *this) { if (pred(val)) return &val; } return nullptr; }
```
It uses a technique called SFINAE, (Substitution Failure Is Not An Error), you can look it up if you're interested, and you can even use it to improve `Find(std::string const&)` and `Remove(std::string const&)` such that they don't unnecessarily allocate a new string when you pass it a `char const*`.